src/daft-dsl/src/resolve_expr/mod.rs

```diff
@@ -337,6 +286,55 @@ impl ExprResolver {
             ))),
         }
     }
+
+    fn validate_expr(&self, expr: ExprRef) -> DaftResult<ExprRef> {
+        if has_agg(&expr) {
+            return Err(DaftError::ValueError(format!(
+                "Aggregation expressions are currently only allowed in agg and pivot: {expr}\nIf you would like to have this feature, please see https://github.com/Eventual-Inc/Daft/issues/1979#issue-2170913383",
+            )));
+        }
+
+        Ok(expr)
+    }
+
+    fn validate_expr_in_agg(&self, expr: ExprRef) -> DaftResult<ExprRef> {
+        let converted_expr = convert_udfs_to_map_groups(&expr);
+
+        if !self.is_valid_expr_in_agg(&converted_expr) {
+            return Err(DaftError::ValueError(format!(
+                "Expressions in aggregations must be composed of non-nested aggregation expressions, got {expr}"
+            )));
+        }
+
+        Ok(converted_expr)
+    }
+
+    /// Checks if an expression used in an aggregation is well formed.
+    /// Expressions for aggregations must be in the form (optional) non-agg expr <- [(agg exprs <- non-agg exprs) or literals or group by keys]
+    ///
+    /// # Examples
+    ///
+    /// Allowed:
+    /// - lit("x")
+    /// - sum(col("a"))
+    /// - sum(col("a")) > 0
+    /// - sum(col("a")) - sum(col("b")) > sum(col("c"))
+    /// - sum(col("a")) + col("b") when "b" is a group by key
+    ///
+    /// Not allowed:
+    /// - col("a") when "a" is not a group by key
+    ///     - not an aggregation
+    /// - sum(col("a")) + col("b") when "b" is not a group by key
+    ///     - not all branches are aggregations, literals, or group by keys
+    fn is_valid_expr_in_agg(&self, expr: &ExprRef) -> bool {
+        self.groupby.unwrap().contains(expr)
```

Comment:
i'd add a comment about how `self.groupby.unwrap()` is safe here since it's checked previously on l:247

Reply:
Oops didn't realize you had some other comments. Will make a quick follow-up PR for these